Feature/ampersand node hierarchy

wherget opened the pull request:
Loading SceneNodes into an ampersand model, and traversing the scene node tree.

Not yet clear whether this is the right way to do it, so initiating a PR for the review process.

spidan committed:
basic scene node model for testing hierarchies

spidan committed:
SceneNodeCollection now correctly requires SceneNode, dummy initialize function for testing

spidan committed:
SceneNode requires correct collection now, added additional scenenode props and default values

spidan committed:
Trying to get rid of circular dependency when using sceneNodeCollection to store children of a scene node

spidan committed:
REST path for scene nodes in the scene node collection

spidan committed:
Get sceneNodes from the database by ID, added very basic function trying to traverse the tree and fetch scene nodes on its way

spidan committed:
Removed sceneNodeComponent collection from scene node until its actually used

spidan committed:
Added scenenode collection to editor

spidan committed:
dummy sceneNodeComponent model

spidan committed:
Readded sceneNode model to the collection

spidan committed:
SceneNode now uses an array of indices for storing children, removed SceneNodeCollection from the SceneNodeModel

spidan committed:
Added parsing function to sceneNode to add child node indices to array

spidan committed:
Renamed function, trying to figure out which ids and funtions to use to traverse tree

spidan committed:
Added sceneNodeCollection to app

spidan committed:
Fetching child nodes in success callback to ensure properly fetched parent node

spidan committed:
Removed fetchNodeTree function, as tree fetching now happens in the callback

spidan committed:
Renamed selectNodeBy ID to fetchNodeTree, added app to requirements of sceneNodeCollection for calling fetchNodeTree from callback

spidan committed:
Removed child array, added childNodes as actual (ampersand-)children.

spidan committed:
Pushing childNodes to scenenode children while parsing

spidan committed:
Parsing node name

spidan committed:
Removed sceneNodeCollection from editor, using the app sceneNodeCollection now

spidan committed:
Parsing the remaining scene node properties

spidan committed:
SceneNodeComponent does not require scene nodes

spidan committed:
Very basic collection for scene node components

spidan committed:
Added sceneNodeComponentCollection to sceneNode

spidan committed:
Removed fetchNodeTree from sceneNodeCollection - getOrFetch with the appropriate parsing function seems to be enough to load the tree

wherget commented on compass-js-frontend/src/main/webapp/model/scenenode.js:
```diff
@@ -0,0 +1,72 @@
+		if(!this.childNodes){
+			this.childNodes = new Array();
+		}
+		console.log("Parsing sceneNode "+this.id);
```
leftover printf debugging

wherget commented on compass-js-frontend/src/main/webapp/model/scenenode.js:
```diff
@@ -0,0 +1,72 @@
+		console.log("initializing SceneNode");
+		this.childNodes = new Array();
+	},
+	parse: function(attrs){
```
I'm fairly certain this can be made more concise.
Am I right in assuming that most of this function's purpose is getting the "children" property into the right shape?

wherget commented on compass-js-frontend/src/main/webapp/collection/component-collection.js:
```diff
@@ -0,0 +1,23 @@
+
+var ComponentCollection = AmpersandRestCollection.extend({
+	model: Component,
+	ajaxConfig: {
```
We should refactor out the ever-repeating `ajaxConfig` into a base class of our own. Not necessarily here, but soon-ish.

wherget commented on compass-js-frontend/src/main/webapp/collection/component-collection.js:
```diff
@@ -0,0 +1,23 @@
+			"Accept": "application/json"
+		}
+	},
+	url: Config.getRESTPath("scenenodecomponents/")
```
this can probably also go into a base class, s.t. only the `"scenenodecomponents/"` part can remain.

ebersold commented on compass-js-frontend/src/main/webapp/app.js:
```diff
@@ -23,6 +24,7 @@ app.extend({
 		app.router = new CompassRouter();
 		app.basePath = "/";
 		app.projects = new ProjectCollection();
+		app.scenenodes = new SceneNodeCollection();
```
Should the scenenodes really be a member of the app? I would expect to have a single root node within a scenario.

ebersold committed:
Add some blank lines to improve readability

ebersold committed:
Introduce the compassmodel base class for all models

ebersold committed:
No need to handle those attributes, just return them and let ampersand do its magic

ebersold committed:
remove debug console.log

ebersold committed:
Make the root SceneNode a member of the editorpage

The scenario only holds an id as a reference to the root node

ebersold committed:
Use sync with promise to fetch scenario and root node

ebersold committed:
Do not handle scenenode children for now

ebersold committed:
Introduce CompassCollection

This abstractsd the ajaxConfig, syncWithPromise and a function to only fetch entries which are present in the collection

ebersold committed:
Make the SceneNodeCollection a CompassCollection

ebersold committed:
Remove model attribute from SceneNodeCollection in order to break circular dependency

ebersold committed:
SceneNode stores its children in a SceneNodeCollection

ebersold committed:
On parsing a SceneNode fill the children collection

ebersold committed:
Fetch the scenenode tree recursively (heavily utilizing promises)

ebersold committed:
Remove debug console.log

wherget commented on compass-js-frontend/src/main/webapp/model/project.js:
```diff
@@ -7,11 +7,11 @@
 "use strict";
 
-var AmpersandModel = require("ampersand-model");
+var CompassModel = require("./compassmodel");
```
shouldn't the file be called compass-model.js (as with other class files). Just for consistency, really.

ebersold committed:
Rename file, for consistency reasons only

ebersold committed:
Remove SceneNodeComponent model and collection

not part of the current user story, will be readded at some point in the future

wherget commented on compass-js-frontend/src/main/webapp/ui/page/editor.js:
```diff
@@ -17,20 +18,40 @@ var EditorPage = BasePage.extend({
 		return this;
 	},
-	initUI: function () {
+	initUI: function (rootNode) {
```
this is still a bit of a misnomer, since it does not actually initialize UI at all.
Also it might get called before `render()` is done under odd conditions, so it's not exactly safe to even initialize UI here.

wherget commented on compass-js-frontend/src/main/webapp/ui/page/editor.js:
```diff
@@ -17,20 +18,40 @@ var EditorPage = BasePage.extend({
+	initUI: function (rootNode) {
+		this.root = rootNode;
 		console.log("Editor fetched scenario: " + this.scenario.name);
+		console.log("Editor fetched root node: " + this.root.name);
```
Do we really want these logs here? Maybe trigger an event that signals loading is done? (which in turn can propery initialize UI, if needed?)

wherget commented on compass-js-frontend/src/main/webapp/collection/scenenode-collection.js:
```diff
@@ -0,0 +1,19 @@
+
+var SceneNodeCollection = CompassCollection.extend({
+	url: Config.getRESTPath("scenenodes/")
+	//we do not specifiy the model (SceneNode) in order to avoid a circular dependency
```
can we specify this via the constructor? s.t. the SceneNode object reference remains local to scenenode.js, but this collection still knows its model constructor; thus avoiding circular `require()`

ebersold replied:
Despite the fact that within the ampersand documentation is an example doing something similar it did not work ...

ebersold committed:
Fire an event as soon as the scene tree is fetched

ebersold committed:
Ups, use the right function to trigger events

wherget commented:
:+1: merge ok from my side.
